src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelExplorer.cs

```diff
@@ -0,0 +1,254 @@
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding
+{
+    /// <summary>
+    /// Associates a model object with it's corresponding <see cref="ModelMetadata"/>.
+    /// </summary>
+    public class ModelExplorer
+    {
+        private readonly IModelMetadataProvider _metadataProvider;
+
+        private object _model;
+        private Func<object, object> _modelAccessor;
+        private Type _modelType;
+
+        /// <summary>
+        /// Creates a new <see cref="ModelExplorer"/>.
+        /// </summary>
+        /// <param name="metadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
+        /// <param name="metadata">The <see cref="ModelMetadata"/>.</param>
+        /// <param name="model">The model object. May be <c>null</c>.</param>
+        public ModelExplorer(
+            [NotNull] IModelMetadataProvider metadataProvider, 
+            [NotNull] ModelMetadata metadata, 
+            object model)
+        {
+            _metadataProvider = metadataProvider;
+            Metadata = metadata;
+            _model = model;
+        }
+
+        public ModelExplorer(
+            IModelMetadataProvider metadataProvider,
+            [NotNull] ModelExplorer container,
+            [NotNull] ModelMetadata metadata,
+            Func<object, object> modelAccessor)
+        {
+            _metadataProvider = metadataProvider;
+            Container = container;
+            Metadata = metadata;
+            _modelAccessor = modelAccessor;
+        }
+
+        public ModelExplorer(
+            IModelMetadataProvider metadataProvider,
+            [NotNull] ModelExplorer container,
+            [NotNull] ModelMetadata metadata,
+            object model)
+        {
+            _metadataProvider = metadataProvider;
+            Container = container;
+            Metadata = metadata;
+            _model = model;
+        }
+
+        /// <summary>
+        /// Gets the container <see cref="ModelExplorer"/>.
+        /// </summary>
+        /// <remarks>
+        /// The <see cref="Container"/> will most commonly be set as a result of calling
+        /// <see cref="GetExplorerForProperty(string)"/>. In this case, the returned <see cref="ModelExplorer"/> will
+        /// have it's <see cref="Container"/> set to the instance upon which <see cref="GetExplorerForProperty(string)"/>
+        /// was called.
+        /// 
+        /// This however is not a requirement. The <see cref="Container"/> is informational, and may not
+        /// represent a type that defines the property represented by <see cref="Metadata"/>. This can
+        /// occur when constructing a <see cref="ModelExplorer"/> based on evaluation of a complex
+        /// expression.
+        /// 
+        /// If calling code relies on a parent-child relationship between <see cref="ModelExplorer"/>
+        /// instances, then use <see cref="ModelMetadata.ContainerType"/> to validate this assumption.
+        /// </remarks>
```

Comment:
nit: blank lines in XML docs don't mean anything.  you probably want `<para>...</para>` here